pkg/client/client.go

```diff
@@ -1833,6 +1864,43 @@ func (c *Client) PodCapacity(ctx context.Context) (int, error) {
 	return int(podCapacityTotal), nil
 }
 
+func (c *Client) CreateOrUpdateConsolePlugin(ctx context.Context, plg *consolev1.ConsolePlugin) error {
+
+	conClient := c.osconclient.ConsoleV1().ConsolePlugins()
+	existing, err := conClient.Get(ctx, plg.GetName(), metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		_, err := conClient.Create(ctx, plg, metav1.CreateOptions{})
+		return errors.Wrap(err, "creating ConsolePlugin object failed")
+	}
+	if err != nil {
+		return errors.Wrap(err, "retrieving ConsolePlugin object failed")
+	}
+
+	required := plg.DeepCopy()
+	mergeMetadata(&required.ObjectMeta, existing.ObjectMeta)
+	required.ResourceVersion = existing.ResourceVersion
+
+	_, err = conClient.Update(ctx, required, metav1.UpdateOptions{})
+	return errors.Wrap(err, "updating ConsolePlugin object failed")
+}
+
+func (c *Client) RegisterConsolePlugin(ctx context.Context, name string) error {
+	consoleClient := c.osopclient.OperatorV1().Consoles()
+	console, err := consoleClient.Get(ctx, clusterConsole, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "retrieving console %q failed", clusterConsole)
+	}
+
+	if slices.Contains(console.Spec.Plugins, name) {
+		klog.V(5).Info("console already contains plugin", name)
+		return nil
+	}
+
+	console.Spec.Plugins = append(console.Spec.Plugins, name)
+	_, err = consoleClient.Update(ctx, console, metav1.UpdateOptions{})
```

Comment:
I wonder if we could use `Patch()` instead to be more specific about what we update in the resource? Especially we could use JSON patch to append the item to the plugins array (https://datatracker.ietf.org/doc/html/rfc6902?_gl=1*1xvkbi9*_ga*NjMxMDY2NzQ4LjE2ODA3NzYyMTU.*_ga_4RQQZ3WGE9*MTY4MjUxMDY4My4yLjEuMTY4MjUxMDcwMC40My4wLjA.#appendix-A.2).

Reply:
Would it be okay to address this in separate patch so that the feature itself can be tested?

Reply:
@simonpasquier , I tried the following (hoping this is what you meant)

```go
	jsonPatch := []jsonpatch.Operation{{
		Operation: "add",
		Path:      "/spec/plugins/-",
		Value:     name,
	}}

	patch, err := json.Marshal(jsonPatch)
	if err != nil {
		panic(err)
	}

	_, err = consoleClient.Patch(ctx, clusterConsole, types.JSONPatchType, patch, metav1.PatchOptions{})
```

But this fails with the default `console` spec (below) which doesn't have the `plugins` array defined. It also doesn't really save us from having to `get` and check if the plugin is already added. 

Do you have a better way of doing this? 

```
apiVersion: operator.openshift.io/v1
kind: Console
metadata:
  name: cluster
spec:
  logLevel: Normal
  managementState: Managed
  operatorLogLevel: Normal
```

Reply:
I haven't tested fully but I believe that we'd have to deal with 2 different patches:
* When `.spec.plugins` isn't defined
```json
[{
		"op": "add",
		"path":      "/spec/plugins/",
		"value":     ["monitoring-plugin"],
}]
```

* When `.spec.plugins` is defined
```json
[{
		"op": "add",
		"path":      "/spec/plugins/-",
		"value":     "monitoring-plugin",
}]
```

> It also doesn't really save us from having to get and check if the plugin is already added.

Sure but it should avoid the risk of overwriting the full object while we only care about inserting one item in the list.

Reply:
> Sure but it should avoid the risk of overwriting the full object while we only care about inserting one item in the list.

I have implemented this but I am not sure if this is less error prone that the method used earlier because we don't risk changing the whole object as k8s api-server already prevent us from doing this (occ) ?

 Isn't `get`, `modify`, `update` proven to be working without issues? If the object changes, after `get` IIUC, k8s will reject the change we made based on the `metadata.resourceVersion` and return a `conflict` error asking us to retry

Reply:
> Isn't get, modify, update proven to be working without issues? If the object changes, after get IIUC, k8s will reject the change we made based on the metadata.resourceVersion and return a conflict error asking us to retry

Sure but it's still less elegant and efficient than a patch operation, especially since the resource may be managed by several controllers.